docs: enrich core-concepts with comprehensive UCP protocol overview (#336)
Universal-Commerce-Protocol/ucp: docs/documentation/core-concepts.md

# Core Concepts

The Universal Commerce Protocol (UCP) is an open standard for interoperability
between commerce entities. It defines a common language and functional primitives
so that platforms and businesses can interoperate securely and reliably across
commercial verticals.

This document provides the detailed technical specification for UCP.
For a complete definition of all data models and schemas, see the
[Schema Reference](../specification/reference.md).

!!! note "Terminology"
    Throughout this documentation, **Platform** refers to any entity that *consumes*
    capabilities — an app, an AI agent, a procurement system, or another business.
    **Business** refers to any entity that *exposes* capabilities — a retailer, a
    supplier, a service provider, or any other participant offering commerce
    functionality. These roles are defined by direction of capability flow, not by
    industry vertical, making UCP equally applicable to B2C, B2B, and agent-to-agent
    commerce.

Its primary goal is to enable:

> * **Platforms:** To dynamically discover and consume the capabilities a business exposes.

Reviewer: nit: line wraps (linter not catching?)

Author: Yes, linter passes on long lines but changed to make it consistent.

* **Businesses:** To declare what they offer and how they operate — once — and have any compatible platform discover and use it without bespoke integrations.

> * **Payment & Credential Providers:** To securely hold sensitive user data and issue tokens or credentials on behalf of users, so that platforms and businesses never handle sensitive payment or identity information directly.

Reviewer 2: Opinion: I don't think this explanation fits the goal of UCP for Payment credential providers, it reads more like how businesses and platforms are using them. I also think we overreach a bit saying that platforms and businesses are not allowed to handle sensitive context if they want to.

Possible alternative phrasing, zoomed in on what UCP is empowering for payment & credential providers.

```suggestion
* **Payment & Credential Providers:** To expose their services — tokenization, vaulting, credential issuance — once, and power secure commerce for users across any compatible platform and business.
```

The Universal Commerce Protocol (UCP) is an open standard designed to facilitate
communication and interoperability between diverse commerce entities. In a
fragmented landscape where consumer platforms, business platforms, payment
providers, and identity providers operate on different systems, UCP provides
a standardized common language and functional primitives.

This document provides the detailed technical specification for UCP.
For a complete definition of all data models and schemas, see the
[Schema Reference](../specification/reference.md).

> !!! note "Terminology"

Reviewer: 👍 

nit: needs line wraps

Author: Done.

>     Throughout this documentation, **Platform** (or *consumer platform*) refers to any entity that *consumes* capabilities — an app, an AI agent, a procurement system, or another business. **Business** (or *business platform*) refers to any entity that *exposes* capabilities — a retailer, a supplier, a service provider, or any other participant offering commerce functionality. These roles are defined by direction of capability flow, not by industry vertical, making UCP equally applicable to B2C, B2B, and agent-to-agent commerce. The terms *consumer surface* and *merchant platform* may appear in earlier drafts and community discussions — they are synonymous with Platform and Business respectively.

Reviewer 3: Within /ucp "merchant platform" only appears in https://github.com/Universal-Commerce-Protocol/ucp/discussions/152 and you're scrubbing all usages of "consumer surface" which only appeared in `core-concepts.md`.  You can likely drop the last line on historical terminology unless you need the continuity with internal documentation.

Author: We discussed during our last TC review and decided to pivot towards using client and provider instead of using Platform and Business to be more futureproof. We aligned that a note is still warranted to avoid any confusion and we can remove it in future. I have made a bunch of changes to the file to conform to the new terms. PTAL again.

Author: after getting a quick sense about the changes it will take to switch to the new terminology, I have reverted back to using Platform and Business. I have removed the additional note as that is no longer need.

>     Throughout this documentation, **Platform** (or *consumer platform*) refers to any entity that *consumes* capabilities — an app, an AI agent, a procurement system, or another business. **Business** (or *business platform*) refers to any entity that *exposes* capabilities — a retailer, a supplier, a service provider, or any other participant offering commerce functionality. These roles are defined by direction of capability flow, not by industry vertical, making UCP equally applicable to B2C, B2B, and agent-to-agent commerce. The terms *consumer surface* and *merchant platform* may appear in earlier drafts and community discussions — they are synonymous with Platform and Business respectively.

Reviewer 3: Quick clarifying question on "consumer platform" in line 1. Is "consumer" intended to indicate it "consumes capabilities" or that the Platform is a B2C Platform similar to how you would qualify "Consumer Software"? Does this risk confusion with the later observation that UCP is equally applicable to B2B and agent-to-agent?

Author: It is the former. I make it clear in the lines just below this. Let me know if that makes sense.

Author: Resolving this comment as we pivoted to a different terminology.

Its primary goal is to enable:

> * **Consumer Platforms:** To dynamically discover and consume the capabilities business platform exposes.

Reviewer: We're now reusing platform across consumer and business, which is confusing. I think previous headings were better, but new text is good.

* Platforms: ...
* Businesses: ... 
* Payment & Credential Providers: ...

Reviewer 3: I tend to agree with @igrigorik that we should avoid using "platform" as an optional qualifier for a Business if there is the definitive "Platform" (the consumer of capabilities) that is already established above.

Author: The reason I was planning to move away from just Platforms & Businesses is that for future B2B use cases, it is going to be confusing when one business is a consumer of a service and other business is a provider of service. Many businesses consider themselves "Platforms" so it does not disambiguate clearly.  Take a B2B example of Logistics provider and retailer, both are businesses, so the platform v/s business separation is not clear enough but if we use consumer platform  / business platform, we can highlight the relationship rooted in consumer/provider of capabilities. This helps both B2B, B2C scenarios and in our documentation we can stick to the direction of capability exchanges and make it more generic. Open to more suggestions and thoughts.

Author: Switched to using Platforms and Businesses.

* **Business Platforms:** To declare what they offer and how they operate — once — and have any compatible platform discover and use it without bespoke integrations.

> * **Payment & Credential Providers:** To securely hold sensitive user data and issue tokens or credentials on behalf of users, so that platforms and businesses never handle raw payment or identity information directly.

Reviewer 3: UCP's identity protection story is really about OAuth delegation (the platform gets an access token, never the user's credentials at the business) so perhaps we say that explicitly, instead of borrowing "raw" which is more footed in payments?

Author: We are actively working on supporting other identity schemes, so I would like to avoid anchoring this in just OAuth delegation. It will get out of date very quickly :)

Author: Made some changes, let me know if they address your feedback.

## Key Goals of UCP

> * **Interoperability:** Bridge the gap between consumer platforms, business platforms,

Reviewer: ```suggestion
* **Interoperability:** Bridge the gap between consumer platforms, businesses,
```

Author: Updated the terminology. Resolving this for now.

    and payment ecosystems.
* **Discovery:** Allow consumer platforms to dynamically discover what
    capabilities a business supports (e.g., "Do they support checkout?", "Do they
    support fulfillment options or identity linking?").
* **Security:** Facilitate secure, standards-based (OAuth 2.0, PCI-DSS
    compliant patterns) exchanges of sensitive user and payment data.
* **Agentic Commerce:** Enable AI agents to act on behalf of any principal —

>     a consumer, an enterprise buyer, or an automated system — to execute

Reviewer 3: To distinguish between the dimensions at play, consider reframing this as:

> Enable AI agents to act on behalf of varied principals (an individual, organization, or another agent) and agentic shopping modalities (human-in-the-loop, fully autonomous).

Author: Made some changes. PTAL.

    tasks autonomously.

## Roles & Participants

> UCP defines the interactions between four primary distinct actors, each playing

Reviewer: ```suggestion
UCP defines the interactions between four distinct actors, each playing
```

a specific role in the commerce lifecycle.

> ### Consumer Platform (Application/Agent)

Reviewer: We use "Platform" through the rest of the docs. I agree that "consumer" is helpful, but to avoid ambiguity between "is Platform different from Consumer Platform" let's keep as is? I think the text below explains it clearly.

Author: Moved to Client and Provider terminology.

The platform is any entity that consumes capabilities exposed by a business —
an AI agent, a mobile app, a procurement system, or another business acting
as a capability consumer. It orchestrates the interaction by discovering what
the business supports and invoking the appropriate capabilities on behalf of
its principal (a user, an automated process, or another system).

* **Responsibilities:** Discovering business capabilities via profiles,
    initiating and managing capability sessions, and acting on behalf of its
    principal within the bounds of negotiated capabilities.
* **Examples:** AI Shopping Assistants, Super Apps, Search Engines, B2B Procurement Systems.

### Business

The entity exposing capabilities. In transactional contexts, businesses
act as the **Merchant of Record (MoR)**, retaining financial liability
and ownership of the transaction — though UCP's capability model is not limited
to transactional use cases.

* **Responsibilities:** Publishing a capability profile, declaring supported
    services and extensions, processing capability invocations, and managing

>     the transaction or interaction lifecycle.

Reviewer 3: Between "managing capability sessions" above and "the transaction or interaction lifecycle" here it may be useful to establish standard terminology that captures that (i) capabilities may or may-not be transactional and (ii) capabilities may or may-not be stateful / have an interaction lifecycle.  I think it would allow these explanations to be more precise without hedged language.

For illustration, `catalog.lookup` is non-transactional/stateless, `cart` is non-transactional/stateful, and `checkout` is transactional/stateful.

Author: Made some changes. PTAL.

## Key Goals of UCP

* **Interoperability:** Bridge the gap between platforms, businesses,
    and payment ecosystems.
* **Discovery:** Allow platforms to dynamically discover what
    capabilities a business supports (e.g., "Do they support checkout?", "Do they
    support fulfillment options or identity linking?").
* **Security:** Facilitate secure, standards-based (OAuth 2.0, PCI-DSS
    compliant patterns) exchanges of sensitive user and payment data.

> * **Agentic Commerce:** Enable AI agents to act on behalf of any principals

Reviewer: ```suggestion
* **Agentic Commerce:** Enable AI agents to act on behalf of varied principals
```

Grammar nit: `any principals` reads awkwardly.

UCP revolves around three fundamental constructs that define how entities
interact.

### Capabilities

Capabilities are standalone, independently versioned features that
a business declares it supports. They are the "verbs" of the protocol —
discrete units of functionality that platforms can discover, negotiate, and
invoke.

Each capability is identified by a reverse-domain name (e.g.,
`dev.ucp.shopping.checkout`) and carries a date-based version. Capabilities
are declared in the business's UCP profile at `/.well-known/ucp` and

> repeated in every response so that the platform always knows the active

Reviewer: negotiated and confirmed in every response -- not just repeated.

feature set for a given interaction.

The following are examples of capabilities defined in UCP — see the [Specification](../specification/overview.md) for the authoritative and up-to-date list.

| Capability | Type | Description |
| :--- | :--- | :--- |
| `dev.ucp.shopping.checkout` | Core | Initiates and completes purchase sessions |
| `dev.ucp.shopping.cart` | Core | Pre-checkout cart management |

> | `dev.ucp.shopping.catalog_search` | Core | Search across a business catalog |

Reviewer: ```suggestion
| `dev.ucp.shopping.catalog.search` | Core | Search across a business catalog |
```

> | `dev.ucp.shopping.catalog_lookup` | Core | Retrieve a specific product by ID |

Reviewer: ```suggestion
| `dev.ucp.shopping.catalog.lookup` | Core | Retrieve a specific product by ID |
```

| `dev.ucp.shopping.order` | Core (webhook) | Order lifecycle events |
| `dev.ucp.common.identity_linking` | Common | OAuth-based account linking |

### Extensions

> Extensions are optional modules that augment a base capability. They use the

Reviewer: ```suggestion
Extensions optionally augment a base capability. They use the
```

To avoid introducing yet another "modules" term.

The following are examples of extensions defined in UCP — see the [Specification](../specification/overview.md) for the authoritative and up-to-date list.

| Extension | Extends | Description |
| :--- | :--- | :--- |
| `dev.ucp.shopping.discount` | checkout, cart | Discount codes and promotions |
| `dev.ucp.shopping.fulfillment` | checkout | Shipping and delivery options |
| `dev.ucp.shopping.ap2_mandate` | checkout | Non-repudiable authorization for autonomous commerce |
| `dev.ucp.shopping.buyer_consent` | checkout | Explicit consent capture |

### Services

**Services** group the operations and events for a vertical under a
reverse-domain namespace (e.g., `dev.ucp.shopping`).
A service declares *what* functionality exists for that vertical; transport
bindings declare *how* it is accessed on the wire.

A single service can be accessed via multiple transport bindings:

| Transport | Format | Best For |
| :--- | :--- | :--- |
| **REST** | OpenAPI 3.1.0 | Standard server-to-server integrations |
| **MCP** | OpenRPC | AI agents via Model Context Protocol |
| **A2A** | Agent Card | Agent-to-Agent protocol integrations |

> | **Embedded** | OpenRPC | In-page embedded checkout experiences |

Reviewer: ```suggestion
| **Embedded** | OpenRPC | Embedded integrations |
```

We have checkout and cart, and it's embedded applies to web and native, keep it broad.

A business declares which transport bindings it supports within each service;
platforms pick whichever fits their context — an AI agent may prefer MCP, a
traditional web app may use REST.

> Service namespaces are also UCP's extensibility mechanism for new verticals —

Reviewer: I think we can simplify this.. 

> Service namespaces are also UCP's extensibility mechanism for new verticals — e.g., `dev.ucp.hotels` may be introduced in the future. Businesses opt in by declaring which services they support.

a vertical simply don't participate in it.

## Discovery & Capability Negotiation

UCP uses a profile-based discovery model. Every business publishes a machine-readable
profile at `/.well-known/ucp` that declares which services, capabilities, and
payment handlers they support. Platforms advertise their own profile URL on each
request via the `UCP-Agent` header.

```text
POST /checkout-sessions HTTP/1.1
UCP-Agent: profile="https://agent.example/profiles/shopping-agent.json"
```

This design enables **permissionless onboarding** — any platform with a
discoverable profile can interact with any business without prior registration.
Businesses may additionally establish trust with known platforms through
out-of-band onboarding & verification mechanisms (API keys, OAuth credentials, mTLS certificates).

### Capability Intersection

Capability negotiation follows a **server-selects** architecture. The business
determines the active capabilities by computing the intersection of its own
declared capabilities with those in the platform's profile:

1. **Intersect by name** — Only capabilities both parties declare are candidates.

> 2. **Select version** — For each matched capability, select the highest version

Reviewer: > Select version — For each matched capability, compute the set of versions present in both the business and platform arrays. Select the highest (latest date). If no mutual version exists, exclude the capability.

   present in both arrays. If no common version exists, the capability is excluded.
3. **Prune orphaned extensions** — Extensions whose parent capability is not in
   the intersection are removed. Pruning repeats until stable (handles chains).

The result is a minimal, mutually compatible capability set. Businesses include
the active capabilities in every response so platforms always know which features
apply to a given interaction.

### Profile Structure

Both business and platform profiles share a common base structure — a `ucp`
object declaring protocol version, services, capabilities, and payment handlers,
alongside a `signing_keys` array of JWK public keys. The `ucp` object differs
between the two: the business profile uses a business-specific schema (hosted at
`/.well-known/ucp`), while the platform profile uses a platform-specific schema
(hosted at a URI the platform advertises per-request). This dual-purpose profile
— capabilities *and* keys in a single document — means discovery and
authentication are resolved together.

```json

> {

Reviewer: This example is wrong / missing required fields...

```json
  {
    "ucp": {
      "version": "2026-01-23",
      "services": {
        "dev.ucp.shopping": [
          {
            "version": "2026-01-23",
            "spec": "https://ucp.dev/2026-01-23/specification/overview",
            "transport": "rest",
            "schema": "https://ucp.dev/2026-01-23/services/shopping/rest.openapi.json",
            "endpoint": "https://business.example.com/ucp/v1"
          }
        ]
      },
      "capabilities": {
        "dev.ucp.shopping.checkout": [{
          "version": "2026-01-23",
          "spec": "https://ucp.dev/2026-01-23/specification/checkout",
          "schema": "https://ucp.dev/2026-01-23/schemas/shopping/checkout.json"
        }],
        "dev.ucp.shopping.fulfillment": [{
          "version": "2026-01-23",
          "spec": "https://ucp.dev/2026-01-23/specification/fulfillment",
          "schema": "https://ucp.dev/2026-01-23/schemas/shopping/fulfillment.json",
          "extends": "dev.ucp.shopping.checkout"
        }]
      },
      "payment_handlers": {
        "com.example.processor_tokenizer": [{
          "id": "processor_tokenizer",
          "version": "2026-01-23",
          "spec": "https://example.com/specs/payments/processor_tokenizer",
          "schema": "https://example.com/specs/payments/merchant_tokenizer.json",
          "...": "..."
        }]
      }
    },
    "signing_keys": [{ "kid": "key_2026", "kty": "EC", "crv": "P-256", "alg": "ES256" }]
  }
```

Author: Thanks for catching that. Updated.

  "ucp": {
    "version": "2026-01-23",
    "services": {
      "dev.ucp.shopping": [
        { "version": "2026-01-23", "transport": "rest", "endpoint": "https://business.example.com/ucp/v1" }
      ]
    },
    "capabilities": {
      "dev.ucp.shopping.checkout": [{ "version": "2026-01-23", "schema": "https://ucp.dev/2026-01-23/schemas/shopping/checkout.json" }],
      "dev.ucp.shopping.fulfillment": [{ "version": "2026-01-23", "extends": "dev.ucp.shopping.checkout" }]
    },
    "payment_handlers": {
      "com.example.processor_tokenizer": [{ "id": "tokenizer_1", "version": "2026-01-23" }]
    }
  },
  "signing_keys": [{ "kid": "key_2026", "kty": "EC", "crv": "P-256", "alg": "ES256" }]
}
```

## Namespace Governance

UCP uses reverse-domain naming to embed governance authority directly into
capability and service identifiers. This eliminates the need for a central
registry — domain owners control their own namespace.

```text
{reverse-domain}.{service}.{capability}
```

| Name | Authority | Who governs |
| :--- | :--- | :--- |
| `dev.ucp.shopping.checkout` | ucp.dev | UCP governing body |

> | `com.shopify.checkout.v2` | shopify.com | Shopify |

Reviewer: ```suggestion
| `com.shopify.catalog` | shopify.com | Shopify |
```

Real-world example.

| `com.example.payments.installments` | example.com | example.com |

The `spec` and `schema` URLs declared in a capability must originate from the
namespace authority domain. Platforms **MUST** validate this binding to prevent
spoofed capabilities.

The `dev.ucp.*` namespace is reserved exclusively for capabilities governed by
the UCP Tech Council. Any vendor can define and publish capabilities under their

> own domain — `org.acme.*` — without UCP involvement. Vendor

Reviewer: ```suggestion
own domain — `org.acme.*` — without UCP maintainer approval. Vendor
```

   technical relationship; the business holds API keys and contracts with its PSP.
2. **Platform ↔ Credential Provider (CP)** — The platform acquires a payment
   instrument (token, encrypted payload) directly from the CP, keeping raw
   credentials off the platform-to-business API.
3. **Platform ↔ Business** — The platform submits the opaque token to the
   business, which processes it via its backend PSP integration.

Credentials flow **platform → business** only; businesses **MUST NOT** echo
credentials back in responses.

### Payment Handlers

Payment Handlers are **specifications**, not entities. They define how a
particular payment instrument is acquired and processed. The distinction:

* **Credential Provider (CP) / PSP** — The participant(s) that issue tokens and
  process payments. Depending on the handler, these may be the same entity or
  separate ones (e.g., Google Pay tokenizes; the business's PSP processes).
* **Payment Handler** — The specification that defines the protocol
  (e.g., `com.google.pay`, `dev.shopify.shop_pay`)

The 3-step payment lifecycle:

1. **Negotiation** — The business advertises available payment handlers in its

>    profile and checkout response.

Reviewer: ```suggestion
   profile and checkout response based on contents and negotiated 
   properties of the checkout.
```

2. **Acquisition** — The platform executes the handler's logic to acquire a
   payment instrument (token or encrypted payload) directly from the CP.
3. **Completion** — The platform submits the instrument to the business, which
   charges funds via its PSP integration.

This architecture keeps raw card data off the platform-to-business API,
minimizing PCI scope for the platform and reducing compliance surface for the
overall integration.

## Security & Authentication

### Identity & Key Discovery

UCP profiles serve a dual purpose: declaring capabilities **and** publishing
signing keys. Both parties resolve keys from the same profile document used for
capability negotiation, eliminating a separate key management step.

Key lookup:

1. Obtain the signer's profile URL (from `UCP-Agent` header or `/.well-known/ucp`).
2. Fetch and cache the profile.
3. Match the `keyid` from `Signature-Input` to a `kid` in `signing_keys[]`.
4. Verify the signature using the corresponding public key.

### Authentication Mechanisms

> UCP is compatible with several authentication models:

Reviewer: ```suggestion
UCP supports multiple authentication models:
```

UCP revolves around three fundamental constructs that define how entities
interact.

### Capabilities

Capabilities are standalone, independently versioned features that
a business declares it supports. They are the "verbs" of the protocol —
discrete units of functionality that platforms can discover, negotiate, and
invoke.

Each capability is identified by a reverse-domain name (e.g.,
`dev.ucp.shopping.checkout`) and carries a date-based version. Capabilities
are declared in the business's UCP profile at `/.well-known/ucp` and
negotiated and confirmed in every response so that the platform always knows
the active feature set for a given interaction.

The following are examples of capabilities defined in UCP — see the [Specification](../specification/overview.md) for the authoritative and up-to-date list.

> | Capability | Type | Description |

Reviewer: Drop the `Type` column? The name already captures domain `via dev.ucp.{domain}` and what we  call 'core' today will shift once we introduce new domains. We end up with common and domains-specific.

Author: Makes sense. Updated.

| :--- | :--- | :--- |
| `dev.ucp.shopping.checkout` | Core | Initiates and completes purchase sessions |
| `dev.ucp.shopping.cart` | Core | Pre-checkout cart management |
| `dev.ucp.shopping.catalog.search` | Core | Search across a business catalog |
| `dev.ucp.shopping.catalog.lookup` | Core | Retrieve a specific product by ID |
| `dev.ucp.shopping.order` | Core (webhook) | Order lifecycle events |
| `dev.ucp.common.identity_linking` | Common | OAuth-based account linking |

### Extensions

Extensions optionally augment a base capability. They use the
`extends` field to declare their parent(s) and compose onto the base schema
using JSON Schema `allOf`. Extensions appear in `ucp.capabilities[]` alongside
core capabilities.

```json
{
  "dev.ucp.shopping.fulfillment": [
    {

>       "version": "2026-01-23",

Reviewer: ```suggestion
      "version": "2026-04-08",
```

Stepping back, all the versions + spec links in this draft use 01-23. Do we want to update to 04-28?

Author: updated to 04-08

        "schema": "https://ucp.dev/2026-01-23/schemas/shopping/checkout.json"
      }],
      "dev.ucp.shopping.fulfillment": [{
        "version": "2026-01-23",
        "spec": "https://ucp.dev/2026-01-23/specification/fulfillment",
        "schema": "https://ucp.dev/2026-01-23/schemas/shopping/fulfillment.json",
        "extends": "dev.ucp.shopping.checkout"
      }]
    },
    "payment_handlers": {
      "com.example.processor_tokenizer": [{
        "id": "processor_tokenizer",
        "version": "2026-01-23",
        "spec": "https://example.com/specs/payments/processor_tokenizer",
        "schema": "https://example.com/specs/payments/merchant_tokenizer.json"
      }]
    }
  },
  "signing_keys": [{ "kid": "key_2026", "kty": "EC", "crv": "P-256", "alg": "ES256" }]
}
```

## Namespace Governance

UCP uses reverse-domain naming to embed governance authority directly into
capability and service identifiers. This eliminates the need for a central
registry — domain owners control their own namespace.

```text
{reverse-domain}.{service}.{capability}
```

| Name | Authority | Who governs |
| :--- | :--- | :--- |
| `dev.ucp.shopping.checkout` | ucp.dev | UCP governing body |

> | `com.shopify.catalog` | shopify.com | Shopify |

Reviewer: ```suggestion
| `dev.shopify.catalog` | shopify.com | Shopify |
```

| `com.example.payments.installments` | example.com | example.com |

The `spec` and `schema` URLs declared in a capability must originate from the
namespace authority domain. Platforms **MUST** validate this binding to prevent
spoofed capabilities.

The `dev.ucp.*` namespace is reserved exclusively for capabilities governed by
the UCP Tech Council. Any vendor can define and publish capabilities under their
own domain — `org.acme.*` — without UCP maintainer approval. Vendor
capabilities follow the same extension model, meaning they can extend UCP base
capabilities (e.g., `org.acme.loyalty` extending `dev.ucp.shopping.checkout`)
or define entirely new ones. Because negotiation is always opt-in, vendor
capabilities only activate when both parties declare them, keeping the protocol
decentralized by design.

## Payment Architecture

UCP decouples payment instrument acceptance from payment processing to solve
the N-to-N complexity problem between platforms, businesses, and payment
processors.

### The Trust Triangle

> The payment model is built on three bilateral trust relationships:

Reviewer: Hmm, with new prose we have Business, PSP, CP, Platform.. which is no longer a triangle.

In overview we have:
- Business <> Payment Credential Provider
- Platform <> Payment Credential Provider
- Platform <> Business

We can collapse PSP and CP, or rework this section? We should also make it coherent with content in overview.

Author: Ah! Good catch. I have made it consistent with Overview.

spoofed capabilities.

The `dev.ucp.*` namespace is reserved exclusively for capabilities governed by
the UCP Tech Council. Any vendor can define and publish capabilities under their
own domain — `org.acme.*` — without UCP maintainer approval. Vendor
capabilities follow the same extension model, meaning they can extend UCP base
capabilities (e.g., `org.acme.loyalty` extending `dev.ucp.shopping.checkout`)
or define entirely new ones. Because negotiation is always opt-in, vendor
capabilities only activate when both parties declare them, keeping the protocol
decentralized by design.

## Payment Architecture

UCP decouples payment instrument acceptance from payment processing to solve
the N-to-N complexity problem between platforms, businesses, and payment
processors.

### The Trust Triangle

The payment model is built on three bilateral trust relationships:

1. **Business ↔ Payment Service Provider (PSP)** — Pre-existing legal and
   technical relationship; the business holds API keys and contracts with its PSP.
2. **Platform ↔ Credential Provider (CP)** — The platform acquires a payment
   instrument (token, encrypted payload) directly from the CP, keeping raw

>    credentials off the platform-to-business API.

Reviewer 2: Should this be a may? Its very possible to avoid using credential providers in UCP.

```suggestion
2. **Platform ↔ Credential Provider (CP)** — The platform may leverage an independent credential provider to prepare encrypted or tokenized payloads, keeping raw credentials abstracted away from the platform-to-business APIs.
```

credentials back in responses.

### Payment Handlers

Payment Handlers are **specifications**, not entities. They define how a
particular payment instrument is acquired and processed. The distinction:

* **Credential Provider (CP) / PSP** — The participant(s) that issue tokens and
  process payments. Depending on the handler, these may be the same entity or
  separate ones (e.g., Google Pay tokenizes; the business's PSP processes).
* **Payment Handler** — The specification that defines the protocol
  (e.g., `com.google.pay`, `dev.shopify.shop_pay`)

The 3-step payment lifecycle:

1. **Negotiation** — The business advertises available payment handlers in its
   profile and checkout response based on contents and negotiated
   properties of the checkout.
2. **Acquisition** — The platform executes the handler's logic to acquire a
   payment instrument (token or encrypted payload) directly from the CP.
3. **Completion** — The platform submits the instrument to the business, which
   charges funds via its PSP integration.

This architecture keeps raw card data off the platform-to-business API,
minimizing PCI scope for the platform and reducing compliance surface for the

> overall integration.

Reviewer 2: nit: this mentions cards and PCI, which is bringing in specific instrument context to an otherwise agnostic conversation. I also think that these points have generally been made already, maybe we can just drop this paragraph entirely? Non-blocking.
